File: frontend/src/app/model/ots.ts
export interface OtResponse {
  idOts: number;
  ot: number;
  descripcion: string;
  activo: boolean;
  fechaCreacion: string; // ISO date string
}

export interface OtFullResponse {
  idOts: number;
  ot: number;
  idOtsAnterior?: number | null;

  idCliente?: number | null;
  idArea?: number | null;
  idProyecto?: number | null;
  idFase?: number | null;
  idSite?: number | null;
  idRegion?: number | null;

  descripcion?: string;
  fechaApertura?: string; // 'YYYY-MM-DD'

  idJefaturaClienteSolicitante?: number | null;
  idAnalistaClienteSolicitante?: number | null;
  idCoordinadorTiCw?: number | null;
  idJefaturaResponsable?: number | null;
  idLiquidador?: number | null;
  idEjecutante?: number | null;
  idAnalistaContable?: number | null;

  activo: boolean;
  fechaCreacion: string;
  // NO incluye trabajadores (como lo decidiste)
}

export interface CrearOtCompletaRequest {
  ot: OtCreateRequest;
  trabajadores?: Array<{ idTrabajador: number; rolEnOt: string }>; // opcional
}

export interface OtCreateRequest {
  idOts?: number;               // solo para edición
  idOtsAnterior?: number | null;
  idCliente?: number;
  idArea?: number;
  idProyecto?: number;
  idFase?: number;
  idSite?: number;
  idRegion?: number;
  descripcion?: string;
  fechaApertura?: string;       // 'YYYY-MM-DD'
  idJefaturaClienteSolicitante?: number | null;
  idAnalistaClienteSolicitante?: number | null;
  idCoordinadorTiCw?: number | null;
  idJefaturaResponsable?: number | null;
  idLiquidador?: number | null;
  idEjecutante?: number | null;
  idAnalistaContable?: number | null;
}

export interface Page<T> {
  content: T[];
  pageable: {
    sort: { sorted: boolean; unsorted: boolean; empty: boolean };
    offset: number;
    pageNumber: number;
    pageSize: number;
    paged: boolean;
    unpaged: boolean;
  };
  last: boolean;
  totalPages: number;
  totalElements: number;
  size: number;
  number: number;
  sort: { sorted: boolean; unsorted: boolean; empty: boolean };
  first: boolean;
  numberOfElements: number;
  empty: boolean;
}

File: comfutura/src/main/java/com/backend/comfutura/model/Ots.java
package com.backend.comfutura.model;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;

import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "ots")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Ots {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id_ots")
    private Integer idOts;

    @Column(name = "ot", nullable = false, unique = true)
    private Integer ot;

    @Column(name = "id_ots_anterior")
    private Integer idOtsAnterior;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "id_cliente")
    private Cliente cliente;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "id_area")
    private Area area;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "id_proyecto")
    private Proyecto proyecto;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "id_fase")
    private Fase fase;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id_tipo_ot", nullable = false)
    private TipoOt tipoOt;

    // =========================
    // MODIFICACIÓN AQUÍ: Site y SiteDescripcion
    // =========================

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id_site")
    private Site site;  // Hacer nullable para compatibilidad

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id_site_descripcion")
    private SiteDescripcion siteDescripcion;  // NUEVA RELACIÓN

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "id_region")
    private Region region;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "id_trabajador")
    private Trabajador trabajador;

    @Column(columnDefinition = "TEXT")
    private String descripcion;

    @Column(name = "fecha_apertura", nullable = false)
    private LocalDate fechaApertura;

    // =========================
    // ROLES → TODOS SON TRABAJADORES
    // =========================

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id_jefatura_cliente_solicitante")
    private JefaturaClienteSolicitante jefaturaClienteSolicitante;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id_analista_cliente_solicitante")
    private AnalistaClienteSolicitante analistaClienteSolicitante;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id_coordinador_ti_cw")
    private Trabajador coordinadorTiCw;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id_jefatura_responsable")
    private Trabajador jefaturaResponsable;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id_liquidador")
    private Trabajador liquidador;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id_ejecutante")
    private Trabajador ejecutante;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id_analista_contable")
    private Trabajador analistaContable;

    // =========================
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "id_estado_ot")
    private EstadoOt estadoOt;

    @Column(name = "activo")
    private Boolean activo = true;

    @CreationTimestamp
    @Column(name = "fecha_creacion", updatable = false)
    private LocalDateTime fechaCreacion;

    // =========================
    // MÉTODOS HELPER PARA OBTENER DATOS DE SITE
    // =========================

    @Transient
    public String getSiteDisplayText() {
        if (siteDescripcion != null && siteDescripcion.getSite() != null) {
            String codigo = siteDescripcion.getSite().getCodigoSitio() != null
                    ? siteDescripcion.getSite().getCodigoSitio().trim()
                    : "-";
            String desc = siteDescripcion.getDescripcion() != null
                    ? siteDescripcion.getDescripcion().trim()
                    : "";
            return codigo + " " + desc;
        } else if (site != null) {
            String codigo = site.getCodigoSitio() != null
                    ? site.getCodigoSitio().trim()
                    : "-";
            return codigo;
        }
        return "-";
    }

    @Transient
    public String getSiteCodigo() {
        if (siteDescripcion != null && siteDescripcion.getSite() != null) {
            return siteDescripcion.getSite().getCodigoSitio() != null
                    ? siteDescripcion.getSite().getCodigoSitio().trim()
                    : "-";
        } else if (site != null) {
            return site.getCodigoSitio() != null ? site.getCodigoSitio().trim() : "-";
        }
        return "-";
    }

    @Transient
    public String getSiteDescripcionText() {
        if (siteDescripcion != null) {
            return siteDescripcion.getDescripcion() != null
                    ? siteDescripcion.getDescripcion().trim()
                    : "";
        }
        return "";
    }
}